examples/project_llm_fine_tune/project_llm_fine_fune/assets.py

```diff
@@ -0,0 +1,395 @@
+            from (
+            select
+                book_id,
+                (case when unnest(popular_shelves).name like '%manga%' then 1 else 0 end) as manga
+            from graphic_novels
+            )
+            group by 1
+        );
+    """
+    with duckdb_resource.get_connection() as conn:
+        conn.execute(query)
+
+
+@dg.asset(
+    kinds={"duckdb"},
+    description="Combined book, author and shelf data",
+    group_name="processing",
+)
+def enriched_graphic_novels(
+    duckdb_resource: dg_duckdb.DuckDBResource,
+    graphic_novels,
+    authors,
+    manga_shelf,
+) -> pd.DataFrame:
+    query = f"""
+        create table if not exists enriched_graphic_novels as (
+            select
+            book.title as title,
+            authors.name as author,
+            book.description as description,
+            manga_shelf.is_manga
+            from graphic_novels as book
+            left join authors
+            on book.authors[1].author_id = authors.author_id
+            left join manga_shelf
+            on book.book_id = manga_shelf.book_id
+            where nullif(book.description, '') is not null
+        )
+    """
+    with duckdb_resource.get_connection() as conn:
+        conn.execute(query)
+        select_query = """
+            select * from enriched_graphic_novels
+        """
+        return conn.execute(select_query).fetch_df()
+
+
+def create_prompt_record(data: dict):
+    return {
+        "messages": [
+            {
+                "role": "system",
+                "content": "Given an author, title and book description, is title manga?",
+            },
+            {
+                "role": "user",
+                "content": f"{data['title']} by {data['author']} described as: {data['description']}. Is this a manga?",
+            },
+            {
+                "role": "assistant",
+                "content": data["is_manga"],
+                "weight": 1,
+            },
+        ]
+    }
+
+
+@dg.asset(
+    group_name="preparation",
+    description="Generate training file",
+)
+def training_file(
+    enriched_graphic_novels,
+) -> str:
+    graphic_novels = enriched_graphic_novels.sample(n=constants.TRAINING_FILE_NUM)
+    prompt_data = []
+    for record in [row for _, row in graphic_novels.iterrows()]:
+        prompt_data.append(create_prompt_record(record))
+
+    file_name = "goodreads-training.jsonl"
+    utils.write_openai_file(file_name, prompt_data)
+    return file_name
+
+
+@dg.asset(
+    group_name="preparation",
+    description="Generate validation file",
+)
+def validation_file(
+    enriched_graphic_novels,
+) -> str:
+    graphic_novels = enriched_graphic_novels.sample(n=constants.VALIDATION_FILE_NUM)
+    prompt_data = []
+    for record in [row for _, row in graphic_novels.iterrows()]:
+        prompt_data.append(create_prompt_record(record))
+
+    file_name = "goodreads-validation.jsonl"
```

Comment:
Writing to local files won't work for all deployment types, something to possibly mention in the tutorial.